Reliability improvements for array marshalling

jkoritzinsky opened the pull request:
Guard against byte length overflow. Use a local when marshalling by-ref arrays to avoid multithreading issues.

AaronRobinsonMSFT commented:
@jkoritzinsky Can we get an example of the generated stub?

jkoritzinsky commented:
From the By ref string array test:

```csharp
amespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial void ReverseStrings(ref string[] strArray, out int numElements)
            {
                unsafe
                {
                    ushort **__strArray_gen_native = default;
                    numElements = default;
                    try
                    {
                        //
                        // Marshal
                        //
                        __strArray_gen_native = null;
                        if (strArray != null)
                        {
                            int strArray__bytelen = checked(sizeof(ushort *) * strArray.Length);
                            __strArray_gen_native = (ushort **)System.Runtime.InteropServices.Marshal.AllocCoTaskMem(strArray__bytelen);
                        }

                        {
                            string[] strArray_local = strArray;
                            if (strArray_local != null)
                                for (int __i = 0; __i < strArray_local.Length; ++__i)
                                {
                                    __strArray_gen_native[__i] = null;
                                    if (strArray[__i] != null)
                                    {
                                        __strArray_gen_native[__i] = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(strArray[__i]);
                                    }
                                }
                        }

                        //
                        // Invoke
                        //
                        fixed (int *__numElements_gen_native = &numElements)
                        {
                            ReverseStrings__PInvoke__(&__strArray_gen_native, __numElements_gen_native);
                        }

                        //
                        // Unmarshal
                        //
                        {
                            string[] strArray_local;
                            if (__strArray_gen_native != null)
                            {
                                strArray_local = new string[checked((int)numElements)];
                                for (int __i = 0; __i < strArray_local.Length; ++__i)
                                {
                                    strArray[__i] = __strArray_gen_native[__i] == null ? null : new string ((char *)__strArray_gen_native[__i]);
                                }
                            }
                            else
                                strArray_local = null;
                            strArray = strArray_local;
                        }
                    }
                    finally
                    {
                        //
                        // Cleanup
                        //
                        if (strArray != null)
                            for (int __i = 0; __i < strArray.Length; ++__i)
                            {
                                System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native[__i]);
                            }

                        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native);
                    }
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "reverse_strings")]
            extern private static unsafe void ReverseStrings__PInvoke__(ushort ***strArray, int *numElements);
        }
    }
}
```

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/ArrayMarshallingCodeContext.cs:
```diff
@@ -35,7 +43,17 @@ public ArrayMarshallingCodeContext(Stage currentStage, string indexerIdentifier,
+                info = info with { InstanceIdentifier = info.InstanceIdentifier.Substring(0, info.InstanceIdentifier.Length - LocalManagedIdentifierSuffix.Length) };
+            }
             var (managed, native) = parentContext.GetIdentifiers(info);
+            if (strippedSuffix)
```
I would avoid this branch and update the return to also be the same such as:

```CSharp
var suffix = string.Empty;
...
suffix  = LocalManagedIdentifierSuffix; // On branch
...
return ($"{managed}{suffix}[{indexerIdentifier}]", $"{native}{suffix}[{indexerIdentifier}]");
```

jkotas commented:
```
                        //
                        // Marshal
                        //
                        __strArray_gen_native = null;
```

`__strArray_gen_native = null` is redundant.

jkotas commented:
```
           string[] strArray_local = strArray;
                            if (strArray_local != null)
```
`strArray_local` has to be cached in a local before the buffer allocation. We need to allocate the buffer based on the same array that we will later use for the assignment.

jkotas commented:
```
             __strArray_gen_native[__i] = null;
                                    if (strArray[__i] != null)
                                    {
                                        __strArray_gen_native[__i] = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(strArray[__i]);
                                    }
```
Why can't this be just `__strArray_gen_native[__i] = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(strArray[__i]);` ?

jkotas commented:
> `strArray[__i] = __strArray_gen_native[__i] == null ? null : new string ((char *)__strArray_gen_native[__i]);`

Should this be `strArray_local[__i] = ...` ?

jkotas commented:
```
                        //
                        // Cleanup
                        //
                        if (strArray != null)
                            for (int __i = 0; __i < strArray.Length; ++__i)
```
`strArray` can be changed by the caller by accident at this point. This should use the cached `strArray_local.Length` that matches how much we actually allocated.

jkoritzinsky commented:
Updated codegen:
```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial void ReverseStrings(ref string[] strArray, out int numElements)
            {
                unsafe
                {
                    ushort **__strArray_gen_native = default;
                    numElements = default;
                    //
                    // Setup
                    //
                    string[] strArray_local = strArray;
                    try
                    {
                        //
                        // Marshal
                        //
                        __strArray_gen_native = null;
                        if (strArray != null)
                        {
                            int strArray__bytelen = checked(sizeof(ushort *) * strArray_local.Length);
                            __strArray_gen_native = (ushort **)System.Runtime.InteropServices.Marshal.AllocCoTaskMem(strArray__bytelen);
                        }

                        if (strArray_local != null)
                            for (int __i = 0; __i < strArray_local.Length; ++__i)
                            {
                                __strArray_gen_native[__i] = null;
                                if (strArray_local[__i] != null)
                                {
                                    __strArray_gen_native[__i] = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(strArray_local[__i]);
                                }
                            }

                        //
                        // Invoke
                        //
                        fixed (int *__numElements_gen_native = &numElements)
                        {
                            ReverseStrings__PInvoke__(&__strArray_gen_native, __numElements_gen_native);
                        }

                        //
                        // Unmarshal
                        //
                        if (__strArray_gen_native != null)
                        {
                            strArray_local = new string[checked((int)numElements)];
                            for (int __i = 0; __i < strArray_local.Length; ++__i)
                            {
                                strArray_local[__i] = __strArray_gen_native[__i] == null ? null : new string ((char *)__strArray_gen_native[__i]);
                            }
                        }
                        else
                            strArray_local = null;
                        strArray = strArray_local;
                    }
                    finally
                    {
                        //
                        // Cleanup
                        //
                        if (strArray_local != null)
                            for (int __i = 0; __i < strArray_local.Length; ++__i)
                            {
                                System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native[__i]);
                            }

                        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native);
                    }
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "reverse_strings")]
            extern private static unsafe void ReverseStrings__PInvoke__(ushort ***strArray, int *numElements);
        }
    }
}
```

AaronRobinsonMSFT commented:
I think all uses of `strArray` should be replaced with `strArray_local`. With that done we can collapse some of those `if` branches.

jkoritzinsky commented:
I should be able to replace the rest of the usages of `strArray`, but merging the `if` blocks will be more difficult because they're generated in different places (one in the conditional stackalloc base class and one in the non-blittable array marshaller)

jkotas commented:
I think you still need to have up-front Clear with this structure. If there is an exception thrown by the `System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni` call, the finally block is going to use uninitialized memory.

The alternative that avoids up-front Clear would be to count how many slots got initialized and only free up those in the finally block.

jkoritzinsky commented:
The up front clear is implemented in #379. If you want I can try porting it over to this PR, but I'd rather just get one of the PRs in, fix merge conflicts, and then get the other PR in.

jkoritzinsky commented:
Does anyone have any more comments other than the up-front Clear?

elinor-fung commented on DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs:
```diff
@@ -65,76 +65,100 @@ public override IEnumerable<StatementSyntax> Generate(TypePositionInfo info, Stu
         {
             var (managedIdentifer, nativeIdentifier) = context.GetIdentifiers(info);
 
+            bool cacheManagedValue = info.IsByRef && context.CanUseAdditionalTemporaryState;
```
Maybe pull this out into a helper since `GenerateByteLengthCalculationExpression` and `GenerateNullCheckExpression` use the same logic?

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/Marshalling/ConditionalStackallocMarshallingGenerator.cs:
```diff
@@ -219,6 +208,24 @@ protected bool TryGenerateSetupSyntax(TypePositionInfo info, StubCodeContext con
+                    IdentifierName(context.GetIdentifiers(info).managed),
+                    LiteralExpression(SyntaxKind.NullLiteralExpression));
+        }
+
```
Some extra lines here.

AaronRobinsonMSFT reviewed (approved):
LGTM. My only other ask would be to add any relevant details to our design docs.
